[autograd] enable multithreading autograd

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29574 [autograd] enable multithreading autograd**

Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

We can see that GPU training is not regressed. For the CPU training case, I tested the mnist hogwild traing example on single thread before and after the PR. 

Before the PR: the sampled e2e training times are 195.077s, 204.48s, 193.44s
After the PR: the sampled e2e training times are 200.09s, 195.74s, 190.59s. 

This show cased that single threaded performance is not regressed. 


To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

wanchaol committed:
[autograd] enable multithreading autograd

Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bddppq commented on torch/autograd/gradcheck.py:
```diff
@@ -215,7 +216,10 @@ def gradcheck(func, inputs, eps=1e-6, atol=1e-5, rtol=1e-3, raise_exception=True
-        nondet_tol (float, optional): tolerance for non-determinism. When running
+        nondet_tol (float, optional): absolute tolerance for non-determinism. When running
+            identical inputs through the differentiation, the results must either match
+            exactly (default, 0.0) or be within this tolerance.
```
update the default value in the doc

bddppq commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -202,12 +214,20 @@ Engine::~Engine() {
 
+int Engine::get_num_threads_per_device() { return num_threads_per_device_; }
+
+void Engine::set_num_threads_per_device(int num_threads_per_device) {
```
(discussed offline) Current implementation only takes effect if setting #threads **before** the first run of autograd (which creates the threads). Better to add a check here to raise error if users attempt to set to a different #threads after the thread pool has been created.

bddppq reviewed (commented):
Please paste the benchmark numbers showing that this change doesn't slow down the (default) single thread case.

ezyang commented:
This is a really long diff and the description in the PR is not commensurate with the number of code changes. Please give more information.

ezyang commented:
(If you want an example of what kind of information I'd like to see, check out https://github.com/pytorch/pytorch/pull/28620 )

albanD commented:
I'll wait for this longer description for a full review.
Could you also show in which case this actually improves the performances.

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang reviewed (changes requested):
This PR should NOT be merged without a proper description of what it is doing.

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Tiny non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating pointer number `a+b` is not
   exactly equivalent to `b+a`

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance stays the same.

To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off. For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). Autograd engine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Tiny non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating pointer number `a+b` is not
   exactly equivalent to `b+a`

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off. For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

wanchaol commented:
@ezyang @albanD this should be ready for review now.

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -34,6 +34,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     return {};
 
   auto new_grad = std::move(grads[0]);
+  std::lock_guard<std::mutex> lock(grad_mutex);
```
Above this line there is a comment `// XXX: this method is not thread-safe!` that I'm guessing needs updating

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -34,6 +34,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
   auto new_grad = std::move(grads[0]);
+  std::lock_guard<std::mutex> lock(grad_mutex);
   for (auto& hook : impl::hooks(variable)) {
     new_grad = (*hook)({new_grad})[0];
```
This seems like it could cause a deadlock, if a hook triggers another invocation of backwards, which attempts to reaccumulate into the same variable.

wanchaol replied:
hmm yeah that's true. How about moving the lock after the hooks. The downside of this approach is that the hooks need to retain thread-safe fashion in multithreaded autograd.

ezyang replied:
Yeah... Python hooks are automatically thread safe, but I guess we support C++ hooks too.

albanD replied:
Is that written explicitly in the doc?
Did you checked if we currently have C++ hooks that are not thread safe?

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -858,16 +869,18 @@ auto Engine::start_threads() -> void {
-    std::thread t(&Engine::thread_init, this, i - 1);
-    t.detach();
+  for (int i = 0; i < num_devices; ++i) {
+    for (size_t j = 0; j < num_threads_per_device_; ++j) {
```
I kind of wonder if you actually want a uniform number of threads per cuda device. My understanding from Caffe2's design is that some multithreading over streams on a single gpu device is helpful, but only up to a point. In any case, I'd expect the performance characteristics for CUDA to be different from cPU.

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -277,10 +300,18 @@ auto Engine::thread_main(
-    NodeTask task = queue->pop();
+    std::unique_ptr<NodeTask> task{nullptr};
+    if (graph_task) {
+      task = queue->pop(std::chrono::microseconds(100));
```
What's going on with the timeout here? There doesn't seem to be any explanation of it anywhere.

ezyang commented:
In your benchmark, you only benchmark for CUDA. If I recall correctly, making AccumulateGrad lock free was much more of a pressing issue in the CPU case. Do you have CPU model benchmarks?

ezyang commented on torch/autograd/gradcheck.py:
```diff
@@ -130,7 +130,7 @@ def get_stride(size):
 
 
-def get_analytical_jacobian(input, output, nondet_tol=0.0):
+def get_analytical_jacobian(input, output, nondet_tol=1e-13, nondet_rtol=1e-13):
```
You've modified this test to universally apply nondetermistic tolerance of 1e-13 rather than 0.0. But I don't feel like you should have a large tolerance universally: the nondeterminism ONLY occurs when you run multithreaded autograd. So we should only allow nondeterminism when you have multithreading turned on. It's very nice that autograd is deterministic otherwise (over time we've in fact added features that make it more deterministic), and so let's keep testing for it!

ezyang commented:
> Tiny non-determinism for AccumulateGrad comparing with single thread,
> in multithread environment, because we don't know which child nodes
> will finish computing gradient first, AccumulateGrad will have
> non-deterministic order, and floating pointer number a+b is not
> exactly equivalent to b+a

Two comments:
1. It's not just AccumulateGrad nondeterminism: InputBuffers are like intermediate accumulate grads that occur anywhere in the backwards graph where a forwards result was used multiple times. These are also now nondeterministic with this change. These perturbations probably matter more, because they compound through the rest of the backwards graph.
2. The stated motivation of this change is to achieve better parallelism on compute, right? So I'm wondering if it wouldn't be profitable to pay a little extra memory, and **accumulate into the input buffer / grad deterministically**. The naive solution is to hold onto all of the individual gradients, and then reduce them in parallel in the end. This means we defer the additions to later, but maybe this is an OK price to pay as you're getting most of the wins on the actual compute.

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Tiny non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating pointer number `a+b` is not
   exactly equivalent to `b+a`

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off. For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Tiny non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating pointer number `a+b` is not
   exactly equivalent to `b+a`

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off. For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

albanD commented:
Small comment for the PR description:
- "Tiny non-determinism" it is deterministic or not. Unfortunately, iterations and sgd will make any small difference to be a large difference after a few iterations. "Non-determinism" would be better here.
- Floating point operations are commutative no? They are not associative though, which is the cause of your non-determinism here.

wanchaol reviewed (commented):
> In your benchmark, you only benchmark for CUDA. If I recall correctly, making AccumulateGrad lock free was much more of a pressing issue in the CPU case. Do you have CPU model benchmarks?

The mnist hogwild example benchmark comparison is done on CPU. I can run some other benchmarks with single thread to ensure we are not regressed on the single thread case. 

@ezyang @albanD Thanks for suggestions on how to solve the non-determinism! I think the above approach should work, but when I looked at how to do it correctly, I might need to figure out more details on how to correctly operate on InputBuffer (i.e. how to check the InputBuffer is ready to accumulate, what if two InputBuffers have overlapping variables). Given that the PR is already a big change and non-determinism only happens in multithreaded case, can we incrementally solve this? I will try to fix it in a follow up PR.

wanchaol committed:
merge with master on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

wanchaol commented:
just updated the numbers in the description on the CPU training single threaded case before and after the PR, show cased that the single threaded perf is not regressed.

albanD commented on test/run_test.py:
```diff
@@ -230,10 +231,29 @@ def test_distributed(executable, test_module, test_directory, options):
+    device_count = torch.cuda.device_count() + 1
+    pre_nthreads = os.environ.get('AUTOGRAD_NUM_THREADS_PER_DEVICE')
+    try:
+        for nthreads in [1, min(2, max(1, cpu_count / device_count))]:
```
Does this mean that we are going to run `test_autograd` nb_cpu / (nb_gpu + 1) ? What happens if we have a CPU CI with 128 cores and no GPU? Do we run test_autograd 128 times?

albanD replied:
Ho I misread a range... My bad !

albanD commented on test/test_autograd.py:
```diff
@@ -3073,6 +3074,9 @@ def test_svd_no_singularvectors(self):
             torch.autograd.backward([u, s, v], [torch.ones_like(u), torch.ones_like(s), torch.ones_like(v)])
 
     def test_no_grad_copy(self):
+        if torch.autograd.get_num_threads_per_device() > 1:
```
nit: to be consistent with other skipping mechanism, you might want to do something similar to other skipping mechanisms. For example what we do for multi gpu.
```
@unittest.skipIf(not RUN_CUDA_MULTI_GPU, "multi-GPU not supported")
def test_foo(self):
```

albanD commented on torch/autograd/__init__.py:
```diff
@@ -179,5 +179,17 @@ def variable(*args, **kwargs):
     warnings.warn("torch.autograd.variable(...) is deprecated, use torch.tensor(...) instead")
     return torch.tensor(*args, **kwargs)
 
+# [Experimental] Configure number of threads on each device to concurrently run autograd backward.
```
I think you want to add these as proper docstrings and add the corresponding `.. autofunction:: get_num_threads_per_device ` to `docs/source/autograd.rst`.
See the `.grad` function above for an example.

albanD commented on torch/autograd/__init__.py:
```diff
@@ -179,5 +179,17 @@ def variable(*args, **kwargs):
+
+def set_num_threads_per_device(nthreads):
+    warnings.warn("setting 'num_threads per device' in autograd is currently experimental. "
+                  "It might potentially speed up your backward on CPU but you might notice "
```
nit: I would personally prefer to remove the "might" here and say: "Setting this to a value > 1 will make the backward pass non-deterministic".

albanD commented on torch/autograd/__init__.py:
```diff
@@ -179,5 +179,17 @@ def variable(*args, **kwargs):
+    return Variable._execution_engine.get_num_threads_per_device()
+
+def set_num_threads_per_device(nthreads):
+    warnings.warn("setting 'num_threads per device' in autograd is currently experimental. "
```
You might want to elaborate on why it is experimental? What are the exact limitations and where are they listed explicitly for the user to see them?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -165,7 +165,7 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0):
 
     for jacobian_x, jacobian_reentrant_x in zip(jacobian, jacobian_reentrant):
-        if jacobian_x.numel() != 0 and (jacobian_x - jacobian_reentrant_x).abs().max() > nondet_tol:
+        if not torch.allclose(jacobian_x, jacobian_reentrant_x, nondet_rtol, nondet_tol):
```
Why did you removed the check `jacobian_x.numel() != 0` here?

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -269,14 +273,20 @@ def fn(input):
                     return fail_test('Numerical gradient for function expected to be zero')
         return True
 
+    # multithreaded autograd allow nondeterminism, apply tolerance of 1e-13 rather than 0.0
```
This is the user's responsibility to set these parameters no? If the user set them explicitly as arguments, they will get overwritten here which is not what we want right?

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -46,18 +47,14 @@ static thread_local int worker_device = NO_DEVICE;
 static thread_local int total_depth = 0;
 
+// The timeout that a ReadyQueue pop operation should wait for
+static const int queue_pop_timeout = 10;
```
Wouldn't this be better as a constexpr?

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -179,17 +177,33 @@ auto ReadyQueue::pushShutdownTask() -> void {
   not_empty_.wait(lock, [this]{ return !heap_.empty(); });
   // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
-  auto task = std::move(const_cast<NodeTask&>(heap_.top())); heap_.pop();
+  auto task = make_unique<NodeTask>(std::move(const_cast<NodeTask&>(heap_.top())));
```
Basic cpp question: Is it fine to take a stack allocated object and stuff it in a unique pointer like that (wrt lifetime of the object)?

albanD replied:
So what this line does is:
- The heap_ contains a moved stack object.
- top() returns a const reference
- const_cast removes the const
- move this stack object to the make_unique
- make_unique creates a heap copy of that moved reference?

Is that accurate?

wanchaol replied:
> So what this line does is:

> * make_unique creates a heap copy of that moved reference?
> 
> Is that accurate?

My understanding of [make_unique](https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique) here is that we move the stack object to heap by invoking `NodeTask`s default move constructor. So this object will be in heap instead of stack after the `make_unique`. In the cpp doc it says `make_unique` equivalent to `unique_ptr<T>(new T(std::forward<Args>(args)...))`

wanchaol replied:
@smessmer could you help on giving some insights here? We couldn't be sure about whether this is a valid C++ pattern.

smessmer replied:
Yes, your analysis sounds correct. It calls the move constructor and creates a new NodeTask instance on the heap. Note, however, that the `std::priority_queue` might not expect things being moved out from it. You're violating a class invariant of `std::priority_queue`. It would probably work with most compilers (at least as long as you make sure that the moved-from `NodeTask` object that remains in the queue is still valid to be passed to `CompareNodeTaskTime`), but there might be compilers, systems or standard library implementations where your code breaks.

wanchaol replied:
So this violation of `class invariant of std::priority_queue` seems already exist even before this PR, and the unique_ptr ownership (lifetime) should be fine if we create a new NodeTask instance on the heap with the default move contructor.

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -89,7 +86,8 @@ struct ReadyQueue {
   void push(NodeTask item, bool incrementOutstandingTasks = true);
   void pushShutdownTask();
-  NodeTask pop();
+  std::unique_ptr<NodeTask> pop();
```
Maybe add a comment on why they both exist?

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -277,10 +302,22 @@ auto Engine::thread_main(
-    NodeTask task = queue->pop();
+    std::unique_ptr<NodeTask> task{nullptr};
+    if (graph_task) {
+      // For the case where there're more re-entrant threads than outstanding
```
I'm not sure to understand this comment.
These are not "re-entrant" threads but threads from the pool no?

albanD replied:
Re-entrants don't have to be in a different thread pool. The base behavior is that the current thread gets here.
So the case here is if some other thread in the multithread pool stole the work from this thread, this thread should not block because it needs to keep checking if his work got done by someone else?

albanD commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -289,51 +326,29 @@ auto Engine::thread_main(
+    if (--local_graph_task->outstanding_tasks_ == 0) {
+      // Lock mutex to notify the GraphTask waiting on not_done_
+      std::lock_guard<std::mutex> lock(local_graph_task->mutex_);
+      local_graph_task->not_done_.notify_all();
```
Why do you remove all the logic associated to the case where the owner of the task is a worker thread?

albanD replied:
I think this should be clearly explained somewhere. It's hard to follow that this is correct because of some changes done at some other place in the code.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -38,6 +37,7 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
     new_grad = (*hook)({new_grad})[0];
   }
 
+  std::lock_guard<std::mutex> lock(grad_mutex);
```
This is not needed in the case where there is only 1 thread per device in the autograd engine. But I guess it's not a big performance penalty so not worth the complexity to disable it in that case?

albanD reviewed (changes requested):
Looks better but a few things to check before merging.

Also the CI failures do come from this PR so that should be fixed.

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

We can see that GPU training is not regressed. For the CPU training case, I tested the mnist hogwild traing example on single thread before and after the PR. 

Before the PR: the sampled e2e training times are 195.077s, 204.48s, 193.44s
After the PR: the sampled e2e training times are 200.09s, 195.74s, 190.59s. 

This show cased that single threaded performance is not regressed. 


To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -277,10 +306,23 @@ auto Engine::thread_main(
+      // For the case where there're more re-entrant threads than outstanding
+      // tasks, then some thread that couldn't grab the task will hang on
+      // trying to pop from readyQueue and hold the mutex_, give a timeout to
+      // pop (return null if timeout) to ensure they don't hang forever
```
This still seems wrong to me, on multiple levels.

1. Traditionally, if you hold a lock and then want to do a blocking operation, you're supposed to release the lock before you block (and make sure you have restored all locking invariants when you lock). So this explanation doesn't make sense to me: it seems like you've observed a deadlock... and then added a timeout(!!) of all things to fix the problem. Don't do that!
2.  What if we have run out of tasks in this queue, but we are still making progress on other threads, it seems like you're going to end up with all of the other threads spin-waiting on the empty queue (because they aren't blocking). The degree to which their spinning is made slightly better by however big you make the timeout... but, echoing (1) again, the right way to do the synchronization here is to NOT fix your problem with timeouts.
3. @albanD mentioned to me that you don't have a good way to wake up all the threads when it's time to shut down, which is why you kick them out of blocking every once in a while so they can check for the other condition. This feels self inflicted to me: `condition_variable` explicitly has `notify_one` and `notify_all` for this exact case. You should be able to wake up all the threads when it's time to shut down. If the current API doesn't let you do that, fix the API so you can!

zdevito reviewed (commented):
@ezyang pointed out this line of work to me just now, and I have a number of concerns we should discuss.

Conceptual: PyTorch doesn't have a concept of task-level parallelism like this in any other part of the system.* So this is far from the design space of existing PyTorch where decisions about task-level parallelism is made explicit (either with explicit multi-threading in Python, use of fork/wait primitives, or use of multi-machine RPC). This seems like a pretty drastic change to UX, and warrants discussion at the design level before moving forward.

Flag Proliferation: This introduces flags so it doesn't regress performance. Flags should only exist temporarily because permanent flags split the ecosystem in half and require 2x the amount of work to maintain. What would be the plan to remove this and have it always enabled? I do not see a clear pathway for deciding whether multi-threading the backward will improve or reduce performance, which is concerning. Flags+autoparallelism do not provide a predictable and debuggable user experience. They do not compose well across libraries -- what if one library is improved by this and other is slowed down?

Tech Debt: This increases the complexity of the autograd, a critical component that relatively few developers understand. This will make further changes here harder and its benefits need to be weighted against the extra work needed in the future to make any changes here.

*While the multiple threads in the GPU autograd backward appear to be introducing automatic parallelism, they really are not.  They just make sure we do not have to call setDevice on the same thread repeatedly.

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

We can see that GPU training is not regressed. For the CPU training case, I tested the mnist hogwild traing example on single thread before and after the PR. 

Before the PR: the sampled e2e training times are 195.077s, 204.48s, 193.44s
After the PR: the sampled e2e training times are 200.09s, 195.74s, 190.59s. 

This show cased that single threaded performance is not regressed. 


To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

ezyang commented:
Please comment when you want new review, thanks!

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

We can see that GPU training is not regressed. For the CPU training case, I tested the mnist hogwild traing example on single thread before and after the PR. 

Before the PR: the sampled e2e training times are 195.077s, 204.48s, 193.44s
After the PR: the sampled e2e training times are 200.09s, 195.74s, 190.59s. 

This show cased that single threaded performance is not regressed. 


To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit e797f82:


* **1/1** broken upstream at merge base 457286a (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=457286a38388fad31f80900f9da4b33acbae674c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR
* **1/1** recognized as flaky
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e797f82ab3bd75203ee3a805a7a745a6d3c4169f).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=78017622)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol committed:
Update on "[autograd] enable multithreading autograd"


Summary:

This enable parallel autograd execution on each device which will enable
multithreading autograd (experimental mode). The AutogradEngine will use `num_multithreads_per_device` to create and initialize the threads, and each thread will try to grab available NodeTasks from ReadyQueue for execution. Thread will shutdown when there's no local outstanding task.

There're cases where this could actually improve the performance:
1. training with CPU on multithreading (potentially with JIT to remove
   GIL limitation), backward will not be blocked by single thread with
   this PR.
2. In the case that the autograd graph can be parallel, multithreading
   could parallel graph computation by grabbing tasks concurrently.
3. C++ API for training on CPU, in the case we could do multithreading
   on forward, the backward can be also multithreading with this PR

This is currently set as experimental mode, and it has some constrains:

1. User must set the thread number before the first run of autograd
   where the autograd engine actually create the thread pool. After that
   the option could not be set and we will throw if user doing so.
2. Non-determinism for AccumulateGrad comparing with single thread,
   in multithread environment, because we don't know which child nodes
   will finish computing gradient first, AccumulateGrad will have
   non-deterministic order, and floating point operations are not associative.

To ensure that we don't regression the performance for single-thread
autograd, I ran the fast-rnn benchmark before and after the patch,
result shows that the performance on backward stays the same.
Benchmark setting:
Namespace(cnns=['resnet18', 'resnet50'], device='cuda', group=['cnns', 'rnns'], hiddenSize=512, inputSize=512, miniBatch=64, nloops=100, numLayers=1, print_json=None, rnns=['cudnn', 'aten'], sep=' ', seqLength=100, variable_lstms=False, warmup=10)

Before this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn| 5.405   | 0.1413  | None     | 9.778   | 0.2163  | None
|aten | 8.44    | 0.2018  | None     | 20.78   |1.661    | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.34 |0.211    |None      |33.24    | 0.4363  | None
|resnet50|49.27 |0.7733   |None      |101      | 1.175   | None

After this PR:
Benchmarking LSTMs...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|cudnn|5.315    | 0.5159  | None     | 9.887   | 0.1774  | None
|aten |6.978    |0.2625   |None      | 19.81   | 0.9184  | None

Benchmarking ResNets...

|name | avg_fwd | std_fwd | info_fwd | avg_bwd | std_bwd | info_bwd
|--|--|--|--|--|--|--|
|resnet18|15.41 |0.3535   |None      |33.38    |0.5707   | None
|resnet50|49.11 | 0.5964  |None      |100.7    |1.052    | None

We can see that GPU training is not regressed. For the CPU training case, I tested the mnist hogwild traing example on single thread before and after the PR. 

Before the PR: the sampled e2e training times are 195.077s, 204.48s, 193.44s
After the PR: the sampled e2e training times are 200.09s, 195.74s, 190.59s. 

This show cased that single threaded performance is not regressed. 


To show case that there's performance increase on model training, I set
up the minst hogwild training example with multithreading autograd turn
on and off (on CPU). 

For the case that num_threads = 10 on the whole training
function, we compare num_autograd_threads = 1 v.s.
num_autograd_threads=10, the result shows around 100% improvement on the total training time.

**num_autograd_threads = 1:**
Total training time: 950.6972548961639s 
Test set: Average loss: 0.0341 Accuracy: 9903/10000 (99%)

**num_autograd_threads=10:**
Total training time: 471.45814061164856s
Test set: Average loss: 0.0312 Accuracy: 9902/10000 (99%)


Differential Revision: [D18509157](https://our.internmc.facebook.com/intern/diff/D18509157)

[ghstack-poisoned]

wanchaol commented:
close this in favor of https://github.com/pytorch/pytorch/pull/33157
